### PR TITLE
String indexing via column histograms

### PR DESCRIPTION
**Pull request by Jiaweihu08**
## Description
This PR introduces a new **String** `Transformation` that uses String histograms for value mapping.

### How does it work?
A sorted sequence of distinct string values should be provided to map String values to the space.

The sequence is treated as an equal-width histogram for the column to index.

To transforma a given **String** value, we look for its **insertion position** within the sequence using binary search:
```scala
val hist = ["a", "b", "c", "d", "e"]
val coordinate = hist.search("b").insertionPoint.toDouble / (hist.length - 1) // 0.25
```

Values that are not contained within the limits of the histogram are mapped to their corresponding extremes, i.e. `0.0` or `1.0`. 

### How to use?
To use the feature, we build the histogram and provide it as part of `columnStats` when writing.

`:histogram` should also be specified for the String column to index.

1. Compute the histogram:
- Use all distinct String values:
```scala
import org.apache.spark.sql.DataFrame
import org.apache.spark.sql.functions.col

def getAllString(df: DataFrame, columnName: String): String = {	
  df
  .select(columnName)
  .distinct()
  .na.drop
  .orderBy(col(columnName).asc)
  .collect()
  .map { r => 
    val s = r.getAs[String](0)
    s"'$s'"
  }
  .mkString("[", ",", "]")
}

val histogram = getAllString(df, "test_col_name")
```

- Use a reduced number of String values:
```scala
import org.apache.spark.sql.delta.skipping.MultiDimClusteringFunctions
import org.apache.spark.sql.DataFrame
import org.apache.spark.sql.functions.{col, min}

def getStringHistogramStr(df: DataFrame, columnName: String, numBins: Int): String = {
  val binStarts = "__bin_starts"
  val stringPartitionColumn = MultiDimClusteringFunctions.range_partition_id(col(columnName), numBins)
	
  df
  .select(columnName)
  .distinct()
  .na.drop
  .groupBy(stringPartitionColumn)
  .agg(min(columnName).alias(binStarts))
  .select(binStarts)
  .orderBy(binStarts)
  .collect()
  .map { r => 
    val s = r.getAs[String](0)
    s"'$s'"
  }
  .mkString("[", ",", "]")
}

val histogram = getStringHistogramStr(df, "test_col_name", 100)
```

2. Index the data as follows:
```scala
val columnStats = s"""{"test_col_name_histogram":$histogram}"""

df
  .write
  .format("qbeast")
  .option("columnsToIndex", s"test_col_name:histogram")
  .option("columnStats", columnStats)
  .save(targetPath)
```

If no histogram is provided in `columnStats` during the first write, a default histogram("a" to "z") will be used. Subsequent `appends` will reuse the same histogram.

A new revision will be created when a different custom histogram is provided as "columnStats".

Note: It also fixes the bug where the first write has to have `overwrite` write mode.

## Checklist:

- [x] New feature / bug fix has been committed following the [Contribution guide](https://github.com/Qbeast-io/qbeast-spark/blob/main/CONTRIBUTING.md).
- [x] Add comments to the code (make it easier for the community!).
- [x] Change the documentation.
- [x] Add tests.
- [x] Branch is updated to the main branch.

### CODE REVIEW

File: core/src/main/scala/io/qbeast/core/transform/StringHistogramTransformer.scala [hunk 1]
```diff
@@ -0,0 +1,45 @@
+package io.qbeast.core.transform
+
+import io.qbeast.core.model.QDataType
+import io.qbeast.core.transform.StringHistogramTransformer.{defaultHist, defaultHistStr}
+
+object StringHistogramTransformer extends TransformerType {
+  override def transformerSimpleName: String = "string_hist"
+
+  // "a" to "z"
+  val defaultHist: IndexedSeq[String] = (97 to 122).map(_.toChar.toString)
```
Comment (osopardo1) [inline core/src/main/scala/io/qbeast/core/transform/StringHistogramTransformer.scala:10]: Thinking if the defaultHist could be extracted from the data if the user allows it so...

File: docs/AdvancedConfiguration.md [hunk 2]
```diff
@@ -115,6 +115,57 @@ val columnStats =
...
+2. support for range queries on String columns
+3. improved overall query speed
+
+The following code snippet demonstrates the extraction of a **String** histogram from the source data:
+```scala
+ import org.apache.spark.sql.delta.skipping.MultiDimClusteringFunctions
+ import org.apache.spark.sql.DataFrame
+ import org.apache.spark.sql.functions.{col, min}
+
+ def getStringHistogramStr(columnName: String, numBins: Int, df: DataFrame): String = {
```
Comment (osopardo1) [inline docs/AdvancedConfiguration.md:133]: Maybe not for this PR, but would be nice if the user could call this code or that could activate an option to let qbeast-spark build it internally (for example: spark.qbeast.analyze.histogram.enabled)

what do you think?

Comment (osopardo1) [reply docs/AdvancedConfiguration.md:133]: For next steps.

File: core/src/main/scala/io/qbeast/core/transform/StringHistogramTransformer.scala (hunk 1 above)
Comment (osopardo1) [reply core/src/main/scala/io/qbeast/core/transform/StringHistogramTransformer.scala:10]: For next steps.

**Comment by cdelfosse** [comment]
To merge into the main-1.0.0a branch with Alexey's compaction work.

**Review by osopardo1 — DISMISSED** [review]
All good!

**Comment by osopardo1** [comment]
I am closing this PR, which is replaced by #234 

Since the feature is merged in `main-1.0.0`, we will use the commit there to push it to `main`.

**Pull request closed without merging**